src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,322 @@
+
+        alpha = self.log_alpha.detach().exp()
+
+        q1_pred = self.qf1(torch.Tensor(obs), torch.Tensor(actions))
+        q2_pred = self.qf2(torch.Tensor(obs), torch.Tensor(actions))
+
+        new_next_actions_dist = self.policy(torch.Tensor(next_obs))
+        new_next_actions_pre_tanh, new_next_actions = new_next_actions_dist.rsample(return_pre_tanh_value=True)
+        new_log_pi = new_next_actions_dist.log_prob(value=new_next_actions, pre_tanh_value=new_next_actions_pre_tanh)
+
+        target_q_values = torch.min(
+            self.target_qf1(torch.Tensor(next_obs), new_next_actions),
+            self.target_qf2(torch.Tensor(next_obs), new_next_actions)
+        ).flatten() - (alpha * new_log_pi)
+        with torch.no_grad():
+            q_target = self.reward_scale * torch.Tensor(rewards) + (1. - torch.Tensor(terminals)) * self.discount * target_q_values
+        qf1_loss = F.mse_loss(q1_pred.flatten(), q_target)
+        qf2_loss = F.mse_loss(q2_pred.flatten(), q_target)
+
+        return qf1_loss, qf2_loss
+
+    def update_targets(self):
+        """Update parameters in the target q-functions."""
+        # update for target_qf1
+        target_qfs = [self.target_qf1, self.target_qf2]
+        qfs = [self.qf1, self.qf2]
+        for target_qf, qf in zip(target_qfs, qfs):
+                for t_param, param in zip(target_qf.parameters(),
+                                            qf.parameters()):
+                        t_param.data.copy_(t_param.data * (1.0 - self.tau) +
+                                        param.data * self.tau)
+
+    def optimize_policy(self, itr, gradient_step, samples):
+        """ Optimize the policy based on the policy objective from the sac paper.
+
+        Args:
+            itr (int) - current training iteration
+            samples() - samples recovered from the replay buffer
+        Returns:
+            None
+        """
+
+        obs = samples["observation"]
+
+        # ================update q functions===============#
+        qf1_loss, qf2_loss = self.critic_objective(samples)
+        if(self.gradient_steps == 1):
+            tabular.record("qf_loss/{}".format("qf1_loss"), float(qf1_loss))
+            tabular.record("qf_loss/{}".format("qf2_loss"), float(qf2_loss))
+        elif ((gradient_step % (self.gradient_steps - 1) == 0) and gradient_step != 0):
+            tabular.record("qf_loss/{}".format("qf1_loss"), float(qf1_loss))
+            tabular.record("qf_loss/{}".format("qf2_loss"), float(qf2_loss))
+        self.qf1_optimizer.zero_grad()
+        qf1_loss.backward()
+        # for name, param in self.qf1.named_parameters():
+        #     if param.requires_grad and (param.grad is not None):
+        #         tabular.record("qf1_grads/{}".format(name), float(param.grad.norm()))
+        #         tabular.record("qf2_grads_max/{}".format(name), float(param.grad.max()))
+        #         tabular.record("qf2_grads_min/{}".format(name), float(param.grad.min()))
```

Comment:
By the way, it's often helpful to actually look at the maximum absolute gradient (i.e. `param.grad.abs().max()`), since the actual sign of the gradient is kinda arbitrary.

Also, you should put this commented code in another branch.